src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -27,78 +28,73 @@
 
-    public GiftCertificateDTO getCertificate(long id) {
-        GiftCertificate certificate = certificateRepository.fetchCertificate(id).orElseThrow(() -> new EntityNotFoundException("can't receive GiftCertificate with id = " + id, Error.GiftCertificateNotFound));
-        List<Tag> tags = tagRepository.fetchLinkedTags(id);
-        return certificateDTOMapper.mapToDTO(certificate, tags);
     }
 
     @Transactional
-    public GiftCertificateDTO addCertificate(GiftCertificateDTO certificateDTO) {
-        GiftCertificate giftCertificate = certificateDTOMapper.mapToCertificate(certificateDTO);
-        GiftCertificate certificateFromDb = certificateRepository.createCertificate(giftCertificate);
-        tagRepository.saveTagsForCertificate(certificateDTO.getTags(), certificateFromDb.getId());
-        certificateFromDb = certificateRepository.fetchCertificate(certificateFromDb.getId()).orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST));
-        List<Tag> tagsFromDb = tagRepository.fetchLinkedTags(certificateFromDb.getId());
-        return certificateDTOMapper.mapToDTO(certificateFromDb, tagsFromDb);
+    public GiftCertificate addCertificate(GiftCertificate certificate) {
+        // TODO Is it possible to do it in one row? Without usage of tagsRepo or EntityManager??
+        List<Tag> tags = certificate.getTags();
+        tags = springDataTagRepository.saveAll(tags);
+        tags.stream()
+                .forEach(entityManager::refresh);
+        certificate.setTags(tags);
+        GiftCertificate saved = giftCertificateRepository.save(certificate);
+        entityManager.refresh(saved);
+        return saved;
     }
 
     public void deleteCertificate(long id) {
         certificateRepository.deleteCertificate(id);
     }
 
     @Transactional
-    public boolean updateCertificate(GiftCertificateDTO certificateDTO) {
-        GiftCertificate certificate = certificateDTOMapper.mapToCertificate(certificateDTO);
-
+    public void updateCertificate(GiftCertificate certificate) {
         boolean newTagLinked = false;
-        if (certificateDTO.getTags() != null) {
-            newTagLinked = tagRepository.saveTagsForCertificate(certificateDTO.getTags(), certificate.getId());
+        if (certificate.getTags() != null) {
+            newTagLinked = tagRepository.saveTagsForCertificate(certificate.getTags(), certificate.getId());
         }
-        return (newTagLinked | certificateRepository.updateCertificate(certificateDTO));
+        GiftCertificate certificate1 = giftCertificateRepository.findById(certificate.getId()).get();
+        certificate.setTags(certificate1.getTags());
+        certificate.setCreateDate(certificate1.getCreateDate());
+        certificate.setLastUpdateDate(certificate1.getLastUpdateDate());
+        if(certificate.getDuration()==null) certificate.setDuration(certificate1.getDuration());
```

Comment:
please avoid this type of IFs. use IF with line seps and {}

Reply:
got it. fixed